mm2src/coins/solana/solana_coin.rs

```diff
@@ -40,7 +71,155 @@ impl Deref for SolanaCoin {
+                return MmError::err(SolanaInitError {
+                    ticker,
+                    kind: SolanaInitErrorKind::UnsupportedPrivKeyPolicy {
+                        policy_type: "GlobalHDAccount",
+                    },
+                })
+            },
+            PrivKeyBuildPolicy::WalletConnect { .. } => {
+                return MmError::err(SolanaInitError {
+                    ticker,
+                    kind: SolanaInitErrorKind::UnsupportedPrivKeyPolicy {
+                        policy_type: "WalletConnect",
+                    },
+                })
+            },
+        };
+
+        let keypair = keypair_from_seed(priv_key.as_bytes()).map_to_mm(|e| SolanaInitError {
+            ticker: ticker.clone(),
+            kind: SolanaInitErrorKind::Internal { reason: e.to_string() },
+        })?;
+
+        let address = SolanaAddress::from_str(&keypair.pubkey().to_string()).map_to_mm(|e| SolanaInitError {
+            ticker: ticker.clone(),
+            kind: SolanaInitErrorKind::Internal { reason: e.to_string() },
+        })?;
+
+        let rpc_clients: Vec<Arc<RpcClient>> = nodes.iter().map(|n| Arc::new(RpcClient::new(&n.url))).collect();
+
+        let abortable_system = ctx.abortable_system.create_subsystem().map_to_mm(|e| SolanaInitError {
+            ticker: ticker.clone(),
+            kind: SolanaInitErrorKind::Internal { reason: e.to_string() },
+        })?;
+
+        let fields = SolanaCoinFields {
+            ticker,
+            address,
+            abortable_system,
+            rpc_clients: AsyncMutex::new(rpc_clients),
+            protocol_info,
+            tokens_info: PaMutex::new(HashMap::new()),
+        };
+
+        Ok(SolanaCoin(Arc::new(fields)))
+    }
+
+    pub(crate) async fn rpc_client(&self) -> MmResult<Arc<RpcClient>, String> {
+        let mut rpcs = self.rpc_clients.lock().await;
+
+        if let Some(index) = rpcs.iter().position(|rpc| rpc.get_health().is_ok()) {
+            // Put healthy one to the front.
+            rpcs.rotate_left(index);
+
+            return Ok(rpcs[0].clone());
+        }
+
+        MmError::err("No healthy RPC client found.".to_owned())
+    }
+
+    pub fn add_activated_token(&self, ticker: String, info: super::SolanaTokenProtocolInfo) {
+        self.tokens_info.lock().insert(ticker, info);
+    }
+
+    pub async fn token_balance(
+        &self,
+        token_contract_address: &SolanaAddress,
+    ) -> Result<CoinBalance, MmError<BalanceError>> {
+        let rpc = self
+            .rpc_client()
+            .map_err(|e| BalanceError::Transport(e.into_inner()))
+            .await?;
+
+        if rpc
+            .get_token_accounts_by_owner(&self.address, TokenAccountsFilter::Mint(*token_contract_address))
+            .is_err()
+        {
+            return Ok(CoinBalance {
+                spendable: BigDecimal::zero(),
+                unspendable: BigDecimal::zero(),
+            });
+        };
```

Comment:
Seems strange to me that `get_token_accounts_by_owner` returns an error if there is no token accounts which is equivalent to no balance, are you sure this doesn't return an empty list of token accounts instead?

Reply:
Addresses with no balance fails on `get_token_accounts_by_owner`, that's why we have this gate.